File: largestValueInTreeRow.java
/**
 * Problem: https://leetcode.com/problems/find-largest-value-in-each-tree-row/
 * Approach - BFS
 * TC: O(n)
 * SC: O(n)
 **/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        
        if (root == null) {
            return res;
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            Integer max = null;
            
            for (int i = 0; i < size; ++i) {
                TreeNode num = queue.poll();
                
                if (max == null || num.val > max) {
                    max = num.val;
                }
                
                if (num.left != null) {
                    queue.add(num.left);
                }
                
                if (num.right != null) {
                    queue.add(num.right);
                }
            }
            res.add(max);
        }
        
        return res;
    }
}

File: LargestValueInTreeRow.py
# Time Complexity: O(n) where n is the number of nodes in the tree.
# Space Complexity: O(n) where n is the number of nodes in the tree.
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        result = []
        queue = deque()     # Initialize a queue.
        queue.append(root)
        while queue:        # True if queue has elements in it.
            maxvalue = float('-inf')
            size = len(queue)
            for i in range(size):
                curr_node = queue.popleft()
                if curr_node.val > maxvalue:
                    maxvalue = curr_node.val
                if curr_node.left:      # Check if current node has left child.
                    queue.append(curr_node.left)
                if curr_node.right:     # Check if current node has right child.
                    queue.append(curr_node.right)
            result.append(maxvalue)
        return result
            
        